### PR TITLE
Initial NATS Connector Implementation

### PR DESCRIPTION
**Pull request by johnmccabe**
Raising for visibility, some tweaking/tidy-up required (k8s config, sample producer, multistage build, makefile etc)

Built using `github.com/openfaas-incubator/connector-sdk`.

### CODE REVIEW

**Comment by johnmccabe** [comment]
@alexellis I've reimplemented this using the SDK, I've also pulled out the NATS Streaming option (original implementation let you choose whether to bind to the durable queue rather than plain nats).

- [ ] Can you have a look at the structure to see if you're happy.
- [ ] Can you confirm that you do indeed want NATS streaming pulled out of this connector, or do you want A) a seperate streaming connector, B) make streaming configurable in yaml
- [ ] I've held off raising a PR against the SDK with the openfaas-incubator/kafka-connector#16 change, can you confirm on that issue if you are open to that approach.

**Comment by alexellis** [comment]
Thank you John. Will take a look and get back.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,18 @@
+FROM golang:1.9.2
```
Comment (alexellis) [inline Dockerfile:1]: Probably can do 1.10.x now.

File: config/config.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
+// Copyright (c) OpenFaaS Project 2018. All rights reserved.
```
Comment (alexellis) [inline config/config.go:1]: Author(s) now (feel free to PR any code that this might have been based upon)

File: config/config.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+	return Config{
+		GatewayURL:      gatewayURL,
+		UpstreamTimeout: upstreamTimeout,
+		Topics:          topics,
+		RebuildInterval: rebuildInterval,
+		Broker:          broker,
+		PrintResponse:   printResponse,
+	}
+}
```
Comment (alexellis) [inline config/config.go:74]: How does go.mod etc work with using Go 1.9.2 in the dockerfile? Maybe dep would be better until everything else can move across?

File: nats/broker.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+type broker struct {
+	client *nats.Conn
+}
+
+func NewBroker(config BrokerConfig) *broker {
+	broker := &broker{}
+
+	brokerURL := "nats://" + config.Host + ":4222"
+	for {
+		client, err := nats.Connect(brokerURL, nats.Timeout(config.ConnTimeout), nats.Name(clientName))
```
Comment (alexellis) [inline nats/broker.go:31]: Is any kind of Auth support needed?

File: nats/broker.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+		// check client not nil
+		b.client.QueueSubscribe(topic, queueGroup, func(m *nats.Msg) {
+			log.Printf("Received topic: %s, message: %s", m.Subject, string(m.Data))
+			controller.Invoke(m.Subject, &m.Data)
+		})
+	}
+
+	// interrupt handling
+	wg.Wait()
+	b.client.Close()
```
Comment (alexellis) [inline nats/broker.go:64]: Use defer?

File: nats/broker.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+		time.Sleep(1 * time.Second)
+		// TODO Add healthcheck
+	}
+	return broker
+}
+
+func (b *broker) Subscribe(controller *types.Controller, topics []string) {
+	log.Printf("Configured topics: %v", topics)
+
+	wg := sync.WaitGroup{}
```
Comment (alexellis) [inline nats/broker.go:50]: What's this doing?

**Review by alexellis — COMMENTED** [review]
Added a few thoughts.

**Comment by alexellis** [comment]
@johnmccabe I'll merge so people can start taking a look at this, please check out the comments when you have some time.

**Pull request merged**